keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1794 @@
+
+Beyond this, there are two main mitigations during alpha: conservativism in
+changes to the default behavior, and API opt-in and redeployment required for
+the more aggressive behavior.
+
+The alpha changes to the default backoff curve were chosen because they are
+minimal -- the proposal maintains the existing rate and max cap, and reduces the
+initial value to the point that only introduces 3 excess restarts per pod, the
+first 2 excess in the first 10 seconds and the last excess following in the next
+30 seconds (see [Design Details](#front-loaded-decay-curve-methodology)). For a
+hypothetical node with the max 110 pods all stuck in a simultaneous
+CrashLoopBackoff, API requests to change the state transition would increase at
+its fastest period from ~110 requests/10s to 330 requests/10s. By passing this
+minimal change through the existing SIG-scalability tests, while pursuing manual
+and more detailed periodic benchmarking during the alpha period, we can increase
+the confidence in this change and in the possibility of reducing further in the
+future.
+
+For the `Rapid` case, because the change is more significant, including lowering
+the max cap, there is more risk to node stability expected. This change is of
+interest to be tested in the alpha period by end users, and is why it is still
+included with API opt-in even though the risks are higher. That being said it is
+still a relatively conservative change in an effor to minimize the unknown
+changes for fast feedback during alpha, while improved benchmarking and testing
+occurs. For a hypothetical node with the max 110 pods all stuck in a
+simultaneous `Rapid` CrashLoopBackoff, API requests to change the state
+transition would increase from ~110 requests/10s to 440 requests/10s, and since
+the max cap would be lowered, would exhibit up to 440 requests in excess every
+300s (5 minutes), or an extra 1.4 requests per second once all pods reached
+their max cap backoff. It also should be noted that due to the specifics of the
+configuration required in the Pod manifest, being against an immutable field,
+will require the Pods in question to be redeployed. This means it is unlikely
+that all Pods will be in a simultaneous CrashLoopBackoff even if they are
+designed to quickly crash, since they will all need to be redeployed and
+rescheduled.
+
+## Design Details 
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Front loaded decay curve methodology
+As mentioned above, today the standard backoff curve is a 2x exponential decay
+starting at 10s and capping at 5 minutes, resulting in a composite of the
+standard hockey-stick exponential decay graph followed by a linear rise until
+the heat death of the universe as depicted below:
```

Comment:
It's because of the way we handle backoff:
https://github.com/kubernetes/kubernetes/blob/a7ca13ea29ba5b3c91fd293cdbaec8fb5b30cee2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1336-L1349

So the first time the container exits, there is no backoff delay recorded, but then it adds a backoff key at line 1348.

So the actual (current) backoff implementation is:

- 0 seconds delay for first restart
- 10 seconds for second restart
- `10 * 2^(restart_count - 2)` for subsequent restarts

But those numbers are all delayed up to 10s due to https://github.com/kubernetes/kubernetes/issues/123602

Reply:
Whether or not this is intentional, I don't think we can change it. The difference between a 0 and a 1 second restart is probably negligible in practice, especially under load. So I guess the question is, do we still want to start the backoff at 1 second?

Reply:
Preserved in ee72b80043a1b5b5dc23667a8015d70d494f7e08